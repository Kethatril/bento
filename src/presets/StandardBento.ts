import {
    optimize,
    DefinePlugin,
    Loader,
    NamedModulesPlugin,
    NamedChunksPlugin
} from "webpack";
import { extract } from "extract-text-webpack-plugin";
import { basename, extname } from "path";
import Bento, { Features as BaseFeatures, Env } from "../core/Bento";
import { selector } from "../utils/lang";

export interface Features extends BaseFeatures {
    extractCss: boolean;
    writeManifest: boolean;
}

export default class StandardBento extends Bento {
    /**
     * @protected
     * @type {Features}
     * @memberof StandardBento
     */
    protected features: Features;

    /**
     * @see Bento#configure
     * @protected
     * @memberof StandardBento
     */
    protected configure() {
        // set default addtional features flag
        this.set("extractCss", true).set("writeManifest", true);

        // SCRIPT
        this.addRule(["js", "jsx"], {
            loader: "babel-loader",
            options: {
                presets: [
                    [
                        "@babel/preset-env",
                        {
                            useBuiltIns: "usage"
                        }
                    ]
                ]
            }
        })
            .addPlugin("clean-webpack-plugin", [
                this.config.outputDir,
                {
                    root: this.cwd
                }
            ])
            .addPlugin("webpack-manifest-plugin", (env?: string): any[] => {
                return [
                    {
                        writeToFileEmit: this.features.writeManifest
                    }
                ];
            })
            .addPlugin(NamedChunksPlugin, [
                chunk => {
                    if (chunk.name) {
                        return chunk.name;
                    }

                    return chunk.mapModules(m => {
                        return basename(m.request, extname(m.request));
                    });
                }
            ])
            .addPlugin(NamedModulesPlugin, [])
            .addPlugin(
                optimize.CommonsChunkPlugin,
                [
                    {
                        name: "vendor",
                        minChunks(module) {
                            return /node_modules/.test(module.resource);
                        }
                    }
                ],
                Env.Production
            )
            .addPlugin(
                optimize.CommonsChunkPlugin,
                [
                    {
                        name: "runtime"
                    }
                ],
                Env.Production
            )
            .addPlugin(DefinePlugin, (env?: Env): any[] => {
                return [
                    {
                        "process.env": {
                            NODE_ENV: JSON.stringify(env)
                        }
                    }
                ];
            })

            .addPlugin("uglifyjs-webpack-plugin", [], Env.Production);

        // STYLE
        this.addRule("scss", (env?: string): Loader[] => {
            return extract({
                fallback: {
                    loader: "style-loader",
                    options: {
                        sourceMap: this.features.sourceMap
                    }
                },
                use: [
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: this.features.sourceMap,
                            minimize:
                                env === Env.Production
                                    ? { preset: "default" }
                                    : false
                        }
                    },
                    {
                        loader: "postcss-loader",
                        options: {
                            ident: "postcss",
                            plugins() {
                                const autoprefixer = require("autoprefixer");
                                return [autoprefixer()];
                            },
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "resolve-url-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "sass-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    }
                ]
            });
        })
            .addRule(
                "css",
                [
                    {
                        loader: "style-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    },
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: this.features.sourceMap
                        }
                    }
                ],
                /node_modules/
            )
            .addPlugin("extract-text-webpack-plugin", (env?: string): any[] => {
                const select = selector(env);

                return [
                    {
                        filename: select({
                            default: "[name].css",
                            production: "[name].[contenthash:8].css"
                        }),
                        disable: !(
                            this.features.extractCss && env === Env.Production
                        )
                    }
                ];
            });

        // FONTS & MEDIA
        this.addRule(
            [
                "ttf",
                "eot",
                "woff",
                "woff2",
                "jpeg",
                "jpg",
                "png",
                "gif",
                "svg",
                "mp4",
                "webm"
            ],
            (env?: string): Loader => {
                const select = selector(env);

                return {
                    loader: "file-loader",
                    options: {
                        name: select({
                            default: "[name].[ext]",
                            production: "[name].[hash:8].[ext]"
                        })
                    }
                };
            },
            [this.homeDir, /node_modules/]
        );
    }
}
