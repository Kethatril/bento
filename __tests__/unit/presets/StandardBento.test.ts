import StandardBento from "~/presets/StandardBento";

describe("StandardBento", () => {
    it("works as expected", () => {
        const bento = StandardBento.create({
            homeDir: "app",
            outputDir: "public"
        });

        bento.tinker(manifest => {
            expect(manifest).toMatchSnapshot();
        });
    });
});
