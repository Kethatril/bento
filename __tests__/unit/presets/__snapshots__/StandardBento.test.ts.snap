// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardBento works as expected 1`] = `
Object {
  "entry": Object {},
  "plugins": Object {
    "all": Array [
      Object {
        "args": [Function anonymous],
        "plugin": [Function DefinePlugin],
      },
      Object {
        "args": Array [
          "public",
          Object {
            "root": "<cwd>",
          },
        ],
        "plugin": "clean-webpack-plugin",
      },
      Object {
        "args": [Function anonymous],
        "plugin": "extract-text-webpack-plugin",
      },
    ],
    "production": Array [
      Object {
        "args": Array [
          Object {
            "minChunks": [Function minChunks],
            "name": "vendor",
          },
        ],
        "plugin": [Function CommonsChunkPlugin],
      },
      Object {
        "args": Array [
          Object {
            "name": "manifest",
          },
        ],
        "plugin": [Function CommonsChunkPlugin],
      },
    ],
  },
  "rules": Array [
    Object {
      "ext": Array [
        "js",
        "jsx",
      ],
      "include": Array [
        "<cwd>/app",
      ],
      "loaders": Object {
        "loader": "babel-loader",
        "options": Object {
          "presets": Array [
            Array [
              "@babel/preset-env",
              Object {
                "useBuiltIns": "usage",
              },
            ],
          ],
        },
      },
    },
    Object {
      "ext": Array [
        "scss",
      ],
      "include": Array [
        "<cwd>/app",
      ],
      "loaders": [Function anonymous],
    },
    Object {
      "ext": Array [
        "css",
      ],
      "include": /node_modules/,
      "loaders": Array [
        Object {
          "loader": "style-loader",
          "options": Object {
            "sourceMap": true,
          },
        },
        Object {
          "loader": "css-loader",
          "options": Object {
            "sourceMap": true,
          },
        },
      ],
    },
    Object {
      "ext": Array [
        "ttf",
        "eot",
        "woff",
        "woff2",
        "jpeg",
        "jpg",
        "png",
        "gif",
        "svg",
        "mp4",
        "webm",
      ],
      "include": Array [
        "<cwd>/app",
        /node_modules/,
      ],
      "loaders": [Function anonymous],
    },
  ],
}
`;
